src/core/server/plugins/plugin.ts

```diff
@@ -0,0 +1,159 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import typeDetect from 'type-detect';
+import { ConfigPath } from '../config';
+import { Logger } from '../logging';
+import { PluginInitializerCore, PluginStartCore } from './plugins_core';
+
+/**
+ * Dedicated type for plugin name/id that is supposed to make Map/Set/Arrays
+ * that use it as a key or value more obvious.
+ */
+export type PluginName = string;
+
+/**
+ * Describes the set of required and optional properties plugin can define in its
+ * mandatory JSON manifest file.
+ * @internal
+ */
+export interface PluginManifest {
+  /**
+   * Identifier of the plugin.
+   */
+  readonly id: PluginName;
+
+  /**
+   * Version of the plugin.
+   */
+  readonly version: string;
+
+  /**
+   * The version of Kibana the plugin is compatible with, defaults to "version".
+   */
+  readonly kibanaVersion: string;
+
+  /**
+   * Root configuration path used by the plugin, defaults to "id".
+   */
+  readonly configPath: ConfigPath;
+
+  /**
+   * An optional list of the other plugins that **must be** installed and enabled
+   * for this plugin to function properly.
+   */
+  readonly requiredPlugins: ReadonlyArray<PluginName>;
+
+  /**
+   * An optional list of the other plugins that if installed and enabled **may be**
+   * leveraged by this plugin for some additional functionality but otherwise are
+   * not required for this plugin to work properly.
+   */
+  readonly optionalPlugins: ReadonlyArray<PluginName>;
+
+  /**
+   * Specifies whether plugin includes some client/browser specific functionality
+   * that should be included into client bundle via `public/ui_plugin.js` file.
+   */
+  readonly ui: boolean;
+}
+
+type PluginInitializer<TExposedContract, TDependencies extends Record<PluginName, unknown>> = (
+  core: PluginInitializerCore
+) => {
+  start: (core: PluginStartCore, dependencies: TDependencies) => TExposedContract;
+  stop?: () => void;
+};
+
+/** @internal */
+export class Plugin<
+  TStartContract = unknown,
+  TDependencies extends Record<PluginName, unknown> = Record<PluginName, unknown>
+> {
+  public readonly name: PluginManifest['id'];
+  public readonly configPath: PluginManifest['configPath'];
+  public readonly requiredDependencies: PluginManifest['requiredPlugins'];
+  public readonly optionalDependencies: PluginManifest['optionalPlugins'];
+
+  private instance?: ReturnType<PluginInitializer<TStartContract, TDependencies>>;
+
+  constructor(
+    public readonly path: string,
+    private readonly manifest: PluginManifest,
+    private readonly log: Logger
+  ) {
+    this.name = manifest.id;
+    this.configPath = manifest.configPath;
+    this.requiredDependencies = manifest.requiredPlugins;
+    this.optionalDependencies = manifest.optionalPlugins;
+  }
+
+  public init(core: PluginInitializerCore) {
```

Comment:
__note:__ it was a part of `start` initially, but I've moved it into separate function since we need another arguments for the initializer, we still don't support `init` lifecycle event for plugins though.

Reply:
Can you pass `PluginInitializerCore` to the constructor like you do with `KibanaCore` elsewhere? If so, the logic can go back to being kicked off in `start` (maybe through a `private initializePlugin` function), and we avoid the ambiguity of adding an `init` function to this lightweight wrapper around plugins, especially since we might some day add a real `init` lifecycle handler.

Reply:
Yes, I can, the only thing is that we'll have to create `PluginInitializerCore` even for disabled plugins, but we likely aren't going to have hundreds of disabled plugins so I maybe don't have to worry about that ...

Reply:
Btw, since `KibanaCore` is `BaseServices` now, do you prefer replacing `Core` with `BaseServices` in `PluginInitializerCore` and friends as well for consistency @epixa ?